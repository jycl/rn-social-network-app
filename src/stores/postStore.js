import { observable, computed, toJS, action } from "mobx";
import { getPostHistoryForUser } from "../services/APIService";

/**
 * PostStore is an MobX store that manages the state values
 * related to the users and will update components that are
 * observing it (have this store injected).
 *
 * State params:
 * @param {Array} rawPostList Array of posts for user retrieved from backend API
 * @param {Array} postList list of posts that has been converted from MobX object
 */
class PostStore {
  @observable
  rawPostList = [];

  /**
   * Retrieve user list from backend and save to store
   */
  @action
  loadPostHistory = userId => {
    getPostHistoryForUser(userId).then(list => {
      this.rawPostList = list;
    });
  };

  @computed
  get postList() {
    return toJS(this.rawPostList);
  }
}
const postStore = new PostStore();
export default postStore;
