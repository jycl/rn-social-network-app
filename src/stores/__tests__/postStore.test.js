import postStore from "../postStore";

describe("PostStore tests ", () => {
  it("Snapshot test of PostStore", () => {
    expect(postStore).toMatchSnapshot();
  });

  it("Check initial state values", () => {
    expect(postStore.rawPostList).toHaveLength(0);
    expect(postStore.postList).toHaveLength(0);

    //test set of data
    const postList = [
      {
        id: 1,
        title: "New album out today!"
      },
      {
        id: 2,
        title: "Feeling excited about the latest release..."
      }
    ];
    //set up mock function to load user list and set to rawUserList
    postStore.loadPostHistory = jest.fn().mockImplementation(() => {
      postStore.rawPostList = postList;
    });

    //invoke method and test if raw and filtered list contain correct values
    postStore.loadPostHistory("1");
    expect(postStore.rawPostList).toHaveLength(2);
    expect(postStore.postList).toHaveLength(2);
    expect(postStore.postList[0].title).toBe("New album out today!");
  });

  it("Check current post and post comments", () => {
    expect(postStore.currentPost).toBe(null);
    expect(postStore.currentPostComments).toHaveLength(0);

    //test set of data
    const post = {
      id: 1,
      title: "New album out today!"
    };
    const comments = [
      {
        postId: 1,
        id: 1,
        name: "That's great!",
        body: "Happy for you.",
        name: "John"
      },
      {
        postId: 1,
        id: 2,
        name: "x2!",
        body: "",
        name: "Marie"
      }
    ];

    postStore.loadPostComments = jest.fn().mockImplementation(() => {
      postStore.currentPostComments = comments;
    });
    postStore.setCurrentPost(post);
    postStore.loadPostComments(post.id);

    //invoke method and test if raw and filtered list contain correct values
    expect(postStore.currentPost.title).toBe("New album out today!");
    expect(postStore.currentPostComments).toHaveLength(2);
  });
});
