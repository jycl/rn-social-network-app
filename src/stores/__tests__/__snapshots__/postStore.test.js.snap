// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostStore tests  Snapshot test of PostStore 1`] = `PostStore {}`;
